### ISSUE TITLE
Ltac2: what is or isn't a syntactical value?

### ISSUE THREAD

**Issue by cpitclaudel** [coq/coq#11086]
#### Description of the problem

I've written a function from Ltac2 strings to Coq strings.  It's not very fast, because converting Ltac2 chars to Coq chars is slow.  The fastest implementation I have uses a lookup table: an Ltac2 array of 255 constrs, each corresponding to a character.

This is reasonably fast, but it would be much better if I could compute the table once and for all.  But a restriction on Ltac2 prevents me from doing so:

```coq
Fail Ltac2 number_one := constr:(1).
(* Tactic definition must be a syntactical value *)
```

Similarly:

```coq
Fail Ltac2 number_one := "a".
(* Tactic definition must be a syntactical value *)
```

Why are these two things rejected? 

Thanks!

#### Coq Version

8.9.1

**Comment by ppedrot** [comment]
First one is rejected because it affects the evarmap, second one because it's mutable. Values are recursive applications of constructors.

**Comment by cpitclaudel** [comment]
So, is there a trick to work around this issue ^^? Constructing the lookup table every time is pretty costly.

**Comment by cpitclaudel** [comment]
Also, strings are mutable? :fearful:

**Comment by kyoDralliam** [comment]
I also fail to understand why you cannot bind the result of evaluating a computation to a (syntactic) value.
Wouldn't that solve the `constr:(1)` case ?

**Comment by cpitclaudel** [comment]
@kyoDralliam That would do exactly what I need, actually.

**Comment by ppedrot** [comment]
Nope, terms live in an environment, and that would entail a lot of technical problems in the runtime. We could maybe store the global constraints somewhere, but there are other side-effects that might happen like declaring constants, and this would quickly become a can of worms.

In the current situation, I recommend writing an array thunk, evaluate it once at toplevel and pass it to callees.

**Comment by ppedrot** [comment]
Also, note that mutable strings is not fixed in stone yet. This might be subject to change if people complain loud enough.

**Comment by cpitclaudel** [comment]
🗣️😱 ← here, I did my part ^^

**Comment by cpitclaudel** [comment]
> In the current situation, I recommend writing an array thunk, evaluate it once at toplevel and pass it to callees.

It's a bit tricky.  I use this in a tactic in term, so it gets called independently many times, and I'm not sure how to pass things around.

Can I have a mutable toplevel value that I could stash the value in? Would that have the same problems?

**Comment by ppedrot** [comment]
Yes, a feature that is missing but totally reasonable (and planned) is the ability to set a mutable global Ltac2 binding locally to a computation, that is well-scoped monkey-patching. That would allow this kind of use-case.

**Comment by cpitclaudel** [comment]
Nice, thanks!

**Comment by ppedrot** [comment]
I think the question has been answered, let's close. The runtime replacement wish is implemented in PR #13982, but this issue is not strictly about that.
